env/io_posix.h

```diff
@@ -31,39 +36,81 @@
 #endif
 
 namespace ROCKSDB_NAMESPACE {
-static std::string IOErrorMsg(const std::string& context,
-                              const std::string& file_name) {
-  if (file_name.empty()) {
-    return context;
+std::string IOErrorMsg(const std::string& context,
+                       const std::string& file_name);
+// file_name can be left empty if it is not unkown.
+IOStatus IOError(const std::string& context, const std::string& file_name,
+                 int err_number);
+
+class PosixHelper {
+ public:
+  static size_t GetUniqueIdFromFile(int fd, char* id, size_t max_size);
+  static size_t GetLogicalBufferSize(int fd);
+};
+
+#ifdef OS_LINUX
+// Files under a specific directory have the same logical buffer size.
+// This class caches the logical buffer size for the specified directories to
+// save the CPU cost of computing the size.
+// Safe for concurrent access from multiple threads without any external
+// synchronization.
+class LogicalBufferSizeCache {
+ public:
+  LogicalBufferSizeCache(std::function<size_t(int)> get_logical_buffer_size =
+                             PosixHelper::GetLogicalBufferSize)
+      : get_logical_buffer_size_(get_logical_buffer_size) {}
+
+  // Logical buffer size of files under each specified directory will be cached.
+  // If the logical buffer size for a directory has been cached, the original
+  // cached size will keep being used.
+  void AddCacheDirectories(const std::unordered_set<std::string>& directories) {
+    WriteLock lock(&cache_mutex_);
+    for (const auto& d : directories) {
+      auto& s = cache_[d];
+      (void) s;
+    }
   }
-  return context + ": " + file_name;
-}
 
-// file_name can be left empty if it is not unkown.
-static IOStatus IOError(const std::string& context,
-                        const std::string& file_name, int err_number) {
-  switch (err_number) {
-    case ENOSPC: {
-      IOStatus s = IOStatus::NoSpace(IOErrorMsg(context, file_name),
-                                     strerror(err_number));
-      s.SetRetryable(true);
+  // Returns the logical buffer size for the file.
+  //
+  // 1. The file is under a specified directory. If size is cached, return the
+  // result, otherwise, call GetLogicalBufferSize(fd) and cache the result.
+  // If multiple threads try to call GetLogicalBufferSize concurrently,
+  // only one active thread will do the actual work, other threads will return
+  // when the active thread returns and see the size cached by the active
+  // thread.
+  //
+  // 2. The file is not under any specified directory, the size is always
+  // computed by GetLogicalBufferSize.
+  size_t size(const std::string& fname, int fd) {
+    ReadLock lock(&cache_mutex_);
```

Comment:
I'm a little bit concerned about the potential overhead of the lock. Can you benchmark on a case where max_open_files is much smaller than number of files and see whether it impacts performance?

Reply:
[Here](https://gist.github.com/cheng-chang/11bf2f255c0769dd4890dffdfe7c8e79) is a benchmark to show that the read lock in the cache is not a concern.

Reply:
ran a benchmark, result is [here](https://gist.github.com/cheng-chang/11bf2f255c0769dd4890dffdfe7c8e79).